### PR TITLE
[RFC]uart: Add support for RS485

### PR DESCRIPTION
**Pull request by kumarnac**
Adds support for configuring and enabling RS485 signalling
on supported uarts.

Signed-off-by: Nachiketa Kumar <nachiketa.kumar@intel.com>

### CODE REVIEW

File: drivers/serial/Kconfig [hunk 1]
```diff
@@ -31,6 +31,12 @@ config SERIAL_SUPPORT_INTERRUPT
 	  This is an option to be enabled by individual serial driver
 	  to signal that the driver and hardware supports interrupts.
 
+config SERIAL_SUPPORT_RS_485
+	bool
+	help
+	  This is an option to be enabled by individual serival driver
+	  to signal that the driver and hardware supports RS_455 interface.
```
Comment (talih0) [inline drivers/serial/Kconfig:38]: typo /RS_455/RS_485

File: include/drivers/uart.h [hunk 2]
```diff
@@ -237,6 +241,57 @@ struct uart_event {
...
+typedef enum {
+	/** Full Duplex mode. */
+	UART_RS485_XFER_MODE_FULL_DUPLEX,
+
+	/** Software Controlled Half Duplex mode. */
+	UART_RS485_XFER_MODE_HALF_DUPLEX
+
+} uart_rs485_transfer_mode_t;
+
+/**$
```
Comment (dcpleung) [inline include/drivers/uart.h:258]: Remove extra '$' sign.

**Review by dcpleung — COMMENTED** [review]
Just wondering how complicated it would be if you roll this into the existing `uart_configure()`.

**Review by dcpleung — APPROVED** [review]

File: drivers/serial/Kconfig (hunk 1 above)
Comment (kumarnac) [reply drivers/serial/Kconfig]: Thanks for pointing  out the typo.  Fixed in the latest revision

File: include/drivers/uart.h [hunk 3]
```diff
@@ -237,6 +239,55 @@ struct uart_event {
...
+ * UART RS-485 polarity for de and re signals.
+ */
+typedef enum {
+	/** Polarity for RS485 - Active Low  **/
+	UART_RS485_POL_ACTIVE_LOW = 0,
+
+
+	/** Polarity for RS485 - Active High  **/
+	UART_RS485_POL_ACTIVE_HIGH
+} uart_rs485_polarity_t;
```
Comment (nashif) [inline include/drivers/uart.h:265]: here you call it `rs485`...

File: include/drivers/uart.h [hunk 4]
```diff
@@ -237,6 +239,55 @@ struct uart_event {
...
+
+struct uart_rs_485_config {
```
Comment (nashif) [inline include/drivers/uart.h:267]: and here you call it `_rs_485_`. Can you please use one convention? (rs486 without underscore maybe)

**Review by nashif — CHANGES_REQUESTED** [review]

File: include/drivers/uart.h (hunk 4 above)
Comment (kumarnac) [reply include/drivers/uart.h:267]: Thank for reviewing Anas,  will make the naming uniform.

File: include/drivers/uart.h (hunk 3 above)
Comment (kumarnac) [reply include/drivers/uart.h:264]: Changed in latest vesrion

**Comment by kumarnac** [comment]
Detail on RS485 signalling requirements can be found at https://en.wikipedia.org/wiki/RS-485
This change provides enabling and configuration options for different parameters of RS-485

**Comment by anthonytex** [comment]
May i suggest to include also a general binding like the one present in the Linux kernel to allow for example setting the 485 mode at compile time?
Thanks 
 https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/serial/rs485.yaml

**Comment by galak** [comment]
Is the a driver side or new driver that would utilize RS485 that you can also post?

**Comment by galak** [comment]
want to see how the params here would work in a simpler UART case that uses RTS to the transciever.

**Comment by galak** [comment]
Need to see how we scale the UART api to multiple instances but ones that do an don't support the feature.  (This is a general problem that also exists for async support).

**Comment by carlescufi** [comment]
API meeting:

- Would be great to pull more information from Devicetree
- Does the Kconfig option enable this functionality for all UART instances?
  - This applies to the async option as well, so should be fixed for both

**Comment by kumarnac** [comment]
> Is the a driver side or new driver that would utilize RS485 that you can also post?
The RS485 driver would be a part of a new board with dependencies on external  HAL module and thus not added to this change. 

> Need to see how we scale the UART api to multiple instances but ones that do an don't support the feature. (This is a general problem that also exists for async support).
Explored more on this  and   found that it would be problem for instances that do not support  RS485.




> API meeting:
> 
> * Would be great to pull more information from Devicetree
> * Does the Kconfig option enable this functionality for all UART instances?
>   
>   * This applies to the async option as well, so should be fixed for both

Modified the pull request to include rs485 configuration from the device tree and removed the changes introduced in headers. 
On enabling CONFIG_SERIAL_SUPPORT_RS485 driver implementations can use the uart-rs485-controller.yaml  options to enable instance specific configurations.  Enabling functions would be private implementations let to drivers.

**Comment by kumarnac** [comment]
> May i suggest to include also a general binding like the one present in the Linux kernel to allow for example setting the 485 mode at compile time?
> Thanks
> https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/serial/rs485.yaml

Made change to add uart-rs485-controller.yaml . Can you please take a look.
Thanks

**Comment by alexanderwachter** [comment]
API meeting:

bindings are looking good. Please revisit the types and see whether you can find better ones, like bool to turn on/off something.
@kumarnac mentioned that the Linux bindings are incomplete, so we cannot use them as they are.
@galak, can you check against the Linux driver and give your acknowledgment for not following it exactly?

**Review by anthonytex — COMMENTED** [review]
These are my suggestions.

**Comment by carlescufi** [comment]
@kumarnac will you pick up this PR again to complete the work?
@anthonytex failing the above, would you consider picking it up yourself?

**Comment by anthonytex** [comment]
Hello @carlescufi , Yes i'm going to continue the work if this pr will be closed thanks

**Comment by anthonytex** [comment]
@carlescufi i've opened a new PR, mentioning this please see https://github.com/zephyrproject-rtos/zephyr/pull/40907

**Pull request closed without merging**